internal/ntobjectdir/enum_object_dir.go

```diff
@@ -0,0 +1,111 @@
+package ntobjectdir
+
+import (
+	"path/filepath"
+	"syscall"
+	"unicode/utf16"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/kernel32"
+	"github.com/Microsoft/hcsshim/internal/ntconstants"
+)
+
+//go:generate go run ..\..\mksyscall_windows.go -output zsyscall_windows.go enum_object_dir.go
+
+//sys ntOpenDirectoryObject(handle *uintptr, accessMask uint32, oa *ntconstants.ObjectAttributes) (status uint32) = ntdll.NtOpenDirectoryObject
+//sys ntQueryDirectoryObject(handle uintptr, buffer *byte, length uint32, singleEntry bool, restartScan bool, context *uint32, returnLength *uint32)(status uint32) = ntdll.NtQueryDirectoryObject
+
+const globalNTPath = "\\Global??"
+const bufferSize = 1024
+
+// ntWidePath converts a golang string to a wide string
+func ntWidePath(path string) ([]uint16, error) {
+	path = filepath.Clean(path)
+	fspath := filepath.FromSlash(path)
+	path16 := utf16.Encode(([]rune)(fspath))
+	if len(path16) > 32767 {
+		return nil, syscall.ENAMETOOLONG
+	}
+	return path16, nil
+}
+
+// EnumerateNTGlobalObjectDirectory queries `globalNTPath` for all entires and
+// returns the resulting entry names as a string slice.
+func EnumerateNTGlobalObjectDirectory() ([]string, error) {
+	var (
+		handle uintptr
+		oa     ntconstants.ObjectAttributes
+
+		context      uint32
+		returnLength uint32
+		buffer       [bufferSize]byte
+		result       []string
+	)
+
+	globalNTPath16, err := ntWidePath(globalNTPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// convert `globalNTPath16` to a unicodeString for passing to windows
+	// functions.
+	upathBuffer := kernel32.LocalAlloc(0, int(unsafe.Sizeof(ntconstants.UnicodeString{}))+len(globalNTPath16)*2)
+	defer kernel32.LocalFree(upathBuffer)
+
+	upath := (*ntconstants.UnicodeString)(unsafe.Pointer(upathBuffer))
```

Comment:
I think it would be better to add a `NewUnicodeString(s string) (UnicodeString, error)` and encapsulate this logic there.

Reply:
I updated this but with an additional argument to take an address that can be freed by the caller.